packages/frontend/src/pages/api/voucher/index.ts

```diff
@@ -0,0 +1,218 @@
+import type { NextApiRequest, NextApiResponse } from 'next'
+import { verifyMessage } from 'viem'
+import { getPublicClient } from '@/blockchain/publicClient'
+import { AUCTION_ADDRESSES } from '@/blockchain/auctionAddresses'
+import { AUCTION_ABI } from '@/blockchain/abi/auction'
+import { WinType } from '@/types/winType'
+import { ContractState } from '@/types/ContractState'
+import { environment } from '@/config/environment'
+import * as jose from 'jose'
+import log from '@/utils/log'
+import { buildVoucherClaimMessage } from '@/utils/buildVoucherClaimMessage'
+import { GetVoucherResponse, GetVoucherWithSigRequestSchema } from '@/types/api/voucher'
+import { nonceStore } from '@/utils/nonceStore'
+import { ApiErrorResponse } from '@/types/api/error'
+import { ContractFunctionExecutionError } from 'viem'
+import { getVoucherCodes } from '@/utils/getVoucherCodes'
+
+export default async function handler(req: NextApiRequest, res: NextApiResponse) {
+  switch (req.method) {
+    case 'GET':
+      // GET /api/voucher
+      // Get voucher code with signed JWT
+      return getVoucherWithJwt(req, res)
+    case 'POST':
+      // POST /api/voucher
+      // Submit a signature (of signed nonce) in order to get a voucher code
+      return getVoucherWithSig(req, res)
+    default:
+      res.status(405).json({
+        error: 'Method not allowed',
+      } satisfies ApiErrorResponse)
+  }
+}
+
+async function getVoucherWithJwt(req: NextApiRequest, res: NextApiResponse) {
+  const voucherCodeJwt = req.cookies.voucherCodeJwt
+  if (!voucherCodeJwt) {
+    res.status(403).json({
+      error: 'No JWT supplied',
+    } satisfies GetVoucherResponse)
+    return
+  }
+
+  const { payload } = await jose.jwtVerify(voucherCodeJwt, environment.authSecret, {
+    requiredClaims: ['chainId', 'address'],
+  })
+
+  const winnerIndex = await getWinnerIndex(payload.chainId as number, payload.address as `0x${string}`)
```

Comment:
Let's extract jwt payload interface and use generic instead of `as` syntax as we are passing typed data to `jose.SignJWT`

```typescript
interface JwtPayload {
    chainId: number
    address: Address
}

const { payload } = await jose.jwtVerify<JwtPayload>(voucherCodeJwt, environment.authSecret, {
    requiredClaims: ['chainId', 'address'],
})
```